src/builtins/alias.c

```diff
@@ -0,0 +1,104 @@
+/*
+    psh/backends/posix2/run.c - process-related posix backend
+    Copyright 2020 Manuel Bertele
+
+    This file is part of Psh, P shell.
+
+    Psh is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Psh is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "backend.h"
+#include "builtin.h"
+#include "libpsh/util.h"
+#include "libpsh/xmalloc.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static int add_alias(psh_hash *table, char *alias, char *value)
+{
+
+    psh_hash_add_chk(table, alias, value, 1);
+
+    return 0;
+}
+
+static char *check_for_alias(psh_hash *table, char *alias)
+{
+    return psh_hash_get(table, alias);
+
+}
+
+
+
+int builtin_alias(int argc, char **argv, psh_state *state)
+{
+
+    int i;
+    char *alias, *value, *argstr;
+
+    argstr = strdup(argv[1]);
+
+    /* Combine all argv pointers to one pointer */
+    for (i = 2; i < argc; i++)
+    {
+        xrealloc(argstr, strlen(argstr) + strlen(argv[i]) + 1);
+        strcat(argstr, argv[i]);
+        strcat(argstr, " ");
+    }
+
+    /* Split the string at '=' */
+    for (i = 0; i < strlen(argstr); i++)
+    {
+        if (argstr[i] == '=')
+        {
+            alias = xcalloc(i, sizeof(char));
```

Comment:
Here your size to allocate is incorrect.

Imagine `argstr` is "a=memory", when the `if` clause on line 65 is entered, `i=1`.
However, c-strings requires a terminating '\0', but allocated `alias` can only hold 'a' but not '\0'.
So when someone operates on `alias`, it would simply go past the memory boundary, causing whatever after the address of `a` being printed, or even worse, overwritten.

You should write:
```alias = xcalloc(i+1, sizeof(char));
```

Also, here it's better to use `xmalloc`, since the memory will be overwritten by the string anyway.

When writing C, be extra careful with c-strings and pointers. But don't worry, according to some books, even very experiences programmers can make these errors from time to time.
If in doubt, think of a short string, maybe one or two elements, and test your idea and indexes in your mind or on a piece of paper. If that is still too hard, check out [Address Sanitizer](https://en.wikipedia.org/wiki/AddressSanitizer), compile with it (as simple as adding `-fsanitize=address` to CFLAGS and LDFLAGS, and feed some unusual input to it.